op-plasma/damgr.go

```diff
@@ -2,48 +2,343 @@ package plasma
+				return nil, err
+			}
+			return nil, ErrPendingChallenge
+		}
+	case ChallengeExpired:
+		// challenge was marked as expired, skip
+		return nil, ErrExpiredChallenge
+	case ChallengeResolved:
+		// challenge was resolved, data is available in storage, return directly
+		if !notFound {
+			return data, nil
+		}
+		// data not found in storage, return from challenge resolved input
+		resolvedInput, err := d.state.GetResolvedInput(commitment)
+		if err != nil {
+			return nil, err
+		}
+		return resolvedInput, nil
+	default:
+		if notFound {
+			if d.isExpired(ch.expiresAt) {
+				// we're past the challenge window and the data is not available
+				return nil, ErrMissingPastWindow
+			} else {
+				// continue syncing challenges hoping it eventually is challenged and resolved
+				if err := d.LookAhead(ctx); err != nil {
+					return nil, err
+				}
+				return nil, ErrPendingChallenge
+			}
+		}
+	}
+
+	return data, nil
+}
+
+// isExpired returns whether the given expiration block number is lower or equal to the current head
+func (d *DA) isExpired(bn uint64) bool {
+	return d.origin.Number >= bn
+}
+
+// AdvanceL1Origin syncs any challenge events included in the l1 block, expires any active challenges
+// after the new resolveWindow, computes and signals the new finalized head and sets the l1 block
+// as the new head for tracking challenges. If forwards an error if any new challenge have expired to
+// trigger a derivation reset.
+func (d *DA) AdvanceL1Origin(ctx context.Context, block eth.BlockID) error {
+	// do not repeat for the same origin
+	if block.Number <= d.origin.Number {
+		return nil
+	}
+	// sync challenges for the given block ID
+	if err := d.LoadChallengeEvents(ctx, block); err != nil {
+		return err
+	}
+	// advance challenge window, computing the finalized head
+	bn, err := d.state.ExpireChallenges(block.Number)
 	if err != nil {
-		return Input{}, err
+		return err
+	}
+
+	if bn > d.finalizedHead.Number {
+		ref, err := d.l1.L1BlockRefByNumber(ctx, bn)
+		if err != nil {
+			return err
+		}
+		d.metrics.RecordChallengesHead("finalized", bn)
+
+		// if we get a greater finalized head, signal to the engine queue
+		if d.finalizedHeadSignalFunc != nil {
+			d.finalizedHeadSignalFunc(ctx, ref)
+
+		}
+		// prune old state
+		d.state.Prune(bn)
+		d.finalizedHead = ref
+
+	}
+	d.origin = block
+	d.metrics.RecordChallengesHead("latest", d.origin.Number)
+
+	d.log.Info("processed plasma l1 origin", "origin", block, "next-finalized", bn, "finalized", d.finalizedHead.Number)
+	return nil
+}
+
+// LoadChallengeEvents fetches the l1 block receipts and updates the challenge status
+func (d *DA) LoadChallengeEvents(ctx context.Context, block eth.BlockID) error {
+	//cached with deposits events call so not expensive
+	_, receipts, err := d.l1.FetchReceipts(ctx, block.Hash)
+	if err != nil {
+		return err
+	}
+	d.log.Info("updating challenges", "epoch", block.Number, "numReceipts", len(receipts))
+	for i, rec := range receipts {
```

Comment:
This would be a nice function to split up into smaller components - it's a little hard to read.